README.md
cbor 0.1.0
==========

[![PGXN version](https://badge.fury.io/pg/cbor.svg)](https://badge.fury.io/pg/cbor)
[![Build Status](https://travis-ci.org/paroga/pg-cbor.png)](https://travis-ci.org/paroga/pg-cbor)

This library contains a single PostgreSQL extension, a data type called
`cbor`. It's an implementation of the Concise Binary Object Representation
specified by [RFC 7049](https://tools.ietf.org/html/rfc7049).

Installation
------------

To build it, just do this:

    make
    make install
    make installcheck

If you encounter an error such as:

    "Makefile", line 8: Need an operator

You need to use GNU make, which may well be installed on your system as
`gmake`:

    gmake
    gmake install
    gmake installcheck

If you encounter an error such as:

    make: pg_config: Command not found

Be sure that you have `pg_config` installed and in your path. If you used a
package management system such as RPM to install PostgreSQL, be sure that the
`-devel` package is also installed. If necessary tell the build process where
to find it:

    env PG_CONFIG=/path/to/pg_config make && make installcheck && make install

If you encounter an error such as:

    ERROR:  must be owner of database regression

You need to run the test suite using a super user, such as the default
"postgres" super user:

    make installcheck PGUSER=postgres

Once cbor is installed, you can add it to a database. If you're running
PostgreSQL 9.1.0 or greater, it's a simple as connecting to a database as a
super user and running:

    CREATE EXTENSION cbor;

Dependencies
------------
The `cbor` data type has no dependencies other than PostgreSQL.

Copyright and License
---------------------

Copyright (c) 2010-2018 Patrick Gansterer.

This module is free software; you can redistribute it and/or modify it under
the [PostgreSQL License](http://www.opensource.org/licenses/postgresql).

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose, without fee, and without a written agreement is
hereby granted, provided that the above copyright notice and this paragraph
and the following two paragraphs appear in all copies.

In no event shall Patrick Gansterer be liable to any party for direct,
indirect, special, incidental, or consequential damages, including lost
profits, arising out of the use of this software and its documentation, even
if Patrick Gansterer has been advised of the possibility of such damage.

Patrick Gansterer specifically disclaims any warranties, including, but not
limited to, the implied warranties of merchantability and fitness for a
particular purpose. The software provided hereunder is on an "as is" basis,
and Patrick Gansterer has no obligations to provide maintenance, support,
updates, enhancements, or modifications.


Changes
Revision history for PostgreSQL extension cbor.

0.1.1

0.1.0  2010-10-07 18:31:43
      - Initial version.
      - Implementation in pure PL/pgSQL.
      - Included in [PGXN Manager](https://github.com/theory/pgxn-manager).
      - Not otherwise released.


src/cbor.h
#ifndef __CBOR_H__
#define __CBOR_H__

#include "postgres.h"

typedef uint32 CborEntry;

#define CBORENTRY_POSMASK 0x1FFFFFFF
#define CBORENTRY_TYPEMASK 0xE0000000

#define CBORENTRY_TYPE_UNSIGNEDINTEGER 0x00000000
#define CBORENTRY_TYPE_NEGATIVEINTEGER 0x20000000
#define CBORENTRY_TYPE_BYTESTRING 0x40000000
#define CBORENTRY_TYPE_TEXTSTRING 0x60000000
#define CBORENTRY_TYPE_ARRAY 0x80000000
#define CBORENTRY_TYPE_MAP 0xA0000000
#define CBORENTRY_TYPE_TAG 0xC0000000
#define CBORENTRY_TYPE_FLOATORSIMPLE 0xE0000000

#define CBORENTRY_ENDPOS(ce_, i) ((ce_)[i] & CBORENTRY_POSMASK)
#define CBORENTRY_OFF(ce_, i) ((i) == 0 ? 0 : CBORENTRY_ENDPOS(ce_, i-1))
#define CBORENTRY_VALUE(ce_, i, cnt) ((void*)((char*)((ce_) + cnt) + CBORENTRY_OFF(ce_, i)))

#define CBORENTRY_GETSTR(ce_, i, cnt) (((char*) CBORENTRY_VALUE(ce_, i, cnt)) + VARHDRSZ)
#define CBORENTRY_STRLEN(ce_, i, cnt) (VARSIZE(CBORENTRY_VALUE(ce_, i, cnt)) - VARHDRSZ)
#define CBORENTRY_SETSTRLEN(ce_, i, cnt, len) SET_VARSIZE(CBORENTRY_VALUE(ce_, i, cnt), VARHDRSZ + len)

#define CBOR_SIMPLE_VALUE 0x7FFFFFFFFFFFFF00
#define CBOR_SIMPLEMASK   0xFFFFFFFFFFFFFF00

#define CBORENTRY_INDEFINITE 0x1F
#define CBORENTRY_BREAK 0xFF

typedef struct CborContainer
{
	int32		count;
	CborEntry	entries[1];
}	CborContainer;

typedef struct CborTag
{
	uint64		value;
	CborEntry	entry;
}	CborTag;

typedef struct Cbor
{
	/* varlena header (do not touch directly!) */
	int32		vl_len_;
	CborEntry	root;
}	Cbor;

typedef struct CborValue
{
	CborEntry	type;

	union
	{
		uint64		uint;
		double		flt;
		int32		length;
	}			value;

	struct CborValue *child;
	struct CborValue *next;

}	CborValue;


#define DatumGetCbor(x) ((Cbor*)DatumGetPointer(x))
#define PG_GETARG_CBOR(x)	DatumGetCbor(PG_DETOAST_DATUM(PG_GETARG_DATUM(x)) )
#define PG_RETURN_CBOR(x)	PG_RETURN_POINTER(x)

#endif


src/cbor_op.c
#include "cbor.h"
#include "access/hash.h"

static int	compareCbor(Cbor * a, Cbor * b);
static int	lengthCompareCborText(const struct varlena * a, const struct varlena * b);
static int	cbor_cmp_recursive(CborEntry * a, CborEntry * b, int32 nr, int32 cnt);
static uint32 cbor_hash_recursive(uint32 hash, CborEntry * cbor, int32 nr, int32 cnt);


PG_FUNCTION_INFO_V1(cbor_ne);
Datum
cbor_ne(PG_FUNCTION_ARGS)
{
	Cbor	   *a = PG_GETARG_CBOR(0);
	Cbor	   *b = PG_GETARG_CBOR(1);
	bool		res;

	res = (compareCbor(a, b) != 0);

	PG_FREE_IF_COPY(a, 0);
	PG_FREE_IF_COPY(b, 1);
	PG_RETURN_BOOL(res);
}

PG_FUNCTION_INFO_V1(cbor_lt);
Datum
cbor_lt(PG_FUNCTION_ARGS)
{
	Cbor	   *a = PG_GETARG_CBOR(0);
	Cbor	   *b = PG_GETARG_CBOR(1);
	bool		res;

	res = (compareCbor(a, b) < 0);

	PG_FREE_IF_COPY(a, 0);
	PG_FREE_IF_COPY(b, 1);
	PG_RETURN_BOOL(res);
}

PG_FUNCTION_INFO_V1(cbor_gt);
Datum
cbor_gt(PG_FUNCTION_ARGS)
{
	Cbor	   *a = PG_GETARG_CBOR(0);
	Cbor	   *b = PG_GETARG_CBOR(1);
	bool		res;

	res = (compareCbor(a, b) > 0);

	PG_FREE_IF_COPY(a, 0);
	PG_FREE_IF_COPY(b, 1);
	PG_RETURN_BOOL(res);
}

PG_FUNCTION_INFO_V1(cbor_le);
Datum
cbor_le(PG_FUNCTION_ARGS)
{
	Cbor	   *a = PG_GETARG_CBOR(0);
	Cbor	   *b = PG_GETARG_CBOR(1);
	bool		res;

	res = (compareCbor(a, b) <= 0);

	PG_FREE_IF_COPY(a, 0);
	PG_FREE_IF_COPY(b, 1);
	PG_RETURN_BOOL(res);
}

PG_FUNCTION_INFO_V1(cbor_ge);
Datum
cbor_ge(PG_FUNCTION_ARGS)
{
	Cbor	   *a = PG_GETARG_CBOR(0);
	Cbor	   *b = PG_GETARG_CBOR(1);
	bool		res;

	res = (compareCbor(a, b) >= 0);

	PG_FREE_IF_COPY(a, 0);
	PG_FREE_IF_COPY(b, 1);
	PG_RETURN_BOOL(res);
}

PG_FUNCTION_INFO_V1(cbor_eq);
Datum
cbor_eq(PG_FUNCTION_ARGS)
{
	Cbor	   *a = PG_GETARG_CBOR(0);
	Cbor	   *b = PG_GETARG_CBOR(1);
	bool		res;

	res = (compareCbor(a, b) == 0);

	PG_FREE_IF_COPY(a, 0);
	PG_FREE_IF_COPY(b, 1);
	PG_RETURN_BOOL(res);
}

PG_FUNCTION_INFO_V1(cbor_cmp);
Datum
cbor_cmp(PG_FUNCTION_ARGS)
{
	Cbor	   *a = PG_GETARG_CBOR(0);
	Cbor	   *b = PG_GETARG_CBOR(1);
	int			res;

	res = compareCbor(a, b);

	PG_FREE_IF_COPY(a, 0);
	PG_FREE_IF_COPY(b, 1);
	PG_RETURN_INT32(res);
}


PG_FUNCTION_INFO_V1(cbor_hash);
Datum
cbor_hash(PG_FUNCTION_ARGS)
{
	Cbor	   *cbor = PG_GETARG_CBOR(0);
	uint32		hash;

	hash = cbor_hash_recursive(0, &cbor->root, 0, 1);

	PG_FREE_IF_COPY(cbor, 0);
	PG_RETURN_INT32(hash);
}


static int
compareCbor(Cbor * a, Cbor * b)
{
	return cbor_cmp_recursive(&a->root, &b->root, 0, 1);
}

static int
cbor_cmp_recursive(CborEntry * a, CborEntry * b, int32 nr, int32 cnt)
{
	int32		i;
	uint32		typeA = a[nr] & CBORENTRY_TYPEMASK;
	uint32		typeB = b[nr] & CBORENTRY_TYPEMASK;

	if (typeA < typeB)
		return -1;
	if (typeA > typeB)
		return 1;

	switch (typeA)
	{
		case CBORENTRY_TYPE_UNSIGNEDINTEGER:
		case CBORENTRY_TYPE_NEGATIVEINTEGER:
			{
				uint64	   *valueA = CBORENTRY_VALUE(a, nr, cnt);
				uint64	   *valueB = CBORENTRY_VALUE(b, nr, cnt);

				if (*valueA < *valueB)
					return -1;
				if (*valueA > *valueB)
					return 1;
				break;
			}

		case CBORENTRY_TYPE_BYTESTRING:
		case CBORENTRY_TYPE_TEXTSTRING:
			{
				return lengthCompareCborText(CBORENTRY_VALUE(a, nr, cnt), CBORENTRY_VALUE(b, nr, cnt));
			}

		case CBORENTRY_TYPE_ARRAY:
			{
				CborContainer *valueA = CBORENTRY_VALUE(a, nr, cnt);
				CborContainer *valueB = CBORENTRY_VALUE(b, nr, cnt);

				if (valueA->count < valueB->count)
					return -1;
				if (valueA->count > valueB->count)
					return 1;
				for (i = 0; i < valueA->count; ++i)
				{
					int			tmp = cbor_cmp_recursive(a, b, i, valueA->count);

					if (tmp)
						return tmp;
				}
				break;
			}

		case CBORENTRY_TYPE_MAP:
			{
				CborContainer *valueA = CBORENTRY_VALUE(a, nr, cnt);
				CborContainer *valueB = CBORENTRY_VALUE(b, nr, cnt);

				if (valueA->count < valueB->count)
					return -1;
				if (valueA->count > valueB->count)
					return 1;
				for (i = 0; i < valueA->count * 2; ++i)
				{
					int			tmp = cbor_cmp_recursive(a, b, i, valueA->count * 2);

					if (tmp)
						return tmp;
				}
				break;
			}

		case CBORENTRY_TYPE_TAG:
			{
				CborTag    *valueA = CBORENTRY_VALUE(a, nr, cnt);
				CborTag    *valueB = CBORENTRY_VALUE(b, nr, cnt);

				if (valueA->value < valueB->value)
					return -1;
				if (valueA->value > valueB->value)
					return 1;
				return cbor_cmp_recursive(a, b, 0, 1);
			}

		case CBORENTRY_TYPE_FLOATORSIMPLE:
			{
				uint64	   *valueA = CBORENTRY_VALUE(a, nr, cnt);
				uint64	   *valueB = CBORENTRY_VALUE(b, nr, cnt);

				if ((*valueA & CBOR_SIMPLEMASK) == CBOR_SIMPLE_VALUE)
				{
					uint8		simpleA = *valueA & 0xFF;
					uint8		simpleB = *valueB & 0xFF;

					if ((*valueB & CBOR_SIMPLEMASK) != CBOR_SIMPLE_VALUE)
						return -1;
					if (simpleA < simpleB)
						return -1;
					if (simpleA > simpleB)
						return 1;
				}
				else
				{
					double		fltA = *((double *) valueA);
					double		fltB = *((double *) valueB);

					if ((*valueB & CBOR_SIMPLEMASK) == CBOR_SIMPLE_VALUE)
						return 1;
					if (fltA < fltB)
						return -1;
					if (fltA > fltB)
						return 1;
				}
				break;
			}
	}

	return 0;
}

static uint32
cbor_hash_recursive(uint32 hash, CborEntry * entry, int32 nr, int32 cnt)
{
	int32		i;
	uint32		type = entry[nr] & CBORENTRY_TYPEMASK;

	hash ^= type;
	hash = (hash << 1) | (hash >> 31);

	switch (type)
	{
		case CBORENTRY_TYPE_UNSIGNEDINTEGER:
		case CBORENTRY_TYPE_NEGATIVEINTEGER:
		case CBORENTRY_TYPE_FLOATORSIMPLE:
			{
				hash ^= DatumGetUInt32(hash_any(CBORENTRY_VALUE(entry, nr, cnt), sizeof(uint64)));
				break;
			}

		case CBORENTRY_TYPE_BYTESTRING:
		case CBORENTRY_TYPE_TEXTSTRING:
			{
				bytea	   *value = CBORENTRY_VALUE(entry, nr, cnt);

				hash ^= DatumGetUInt32(hash_any((unsigned char *) value, VARHDRSZ + VARSIZE(value)));
				break;
			}

		case CBORENTRY_TYPE_ARRAY:
			{
				CborContainer *value = CBORENTRY_VALUE(entry, nr, cnt);

				for (i = 0; i < value->count; ++i)
					hash = cbor_hash_recursive(hash, value->entries, i, value->count);
				break;
			}

		case CBORENTRY_TYPE_MAP:
			{
				CborContainer *value = CBORENTRY_VALUE(entry, nr, cnt);

				for (i = 0; i < value->count * 2; ++i)
					hash = cbor_hash_recursive(hash, value->entries, i, value->count * 2);
				break;
			}

		case CBORENTRY_TYPE_TAG:
			{
				CborTag    *value = CBORENTRY_VALUE(entry, nr, cnt);

				hash ^= DatumGetUInt32(hash_any((unsigned char *) &value->value, sizeof(value->value)));
				hash = cbor_hash_recursive(hash, &value->entry, 0, 1);
				break;
			}
	}

	return hash;
}

static int
lengthCompareCborText(const struct varlena * a, const struct varlena * b)
{
	if (VARSIZE(a) == VARSIZE(b))
		return memcmp(VARDATA(a), VARDATA(b), VARSIZE(a) - VARHDRSZ);
	if (VARSIZE(a) < VARSIZE(b))
		return -1;
	return 1;
}
